### PR TITLE
Implement Slash Command AutoComplete

### PR DESCRIPTION
**Pull request by VelvetThePanda**
# Summary
Adds support for slash-command auto-complete. 

# Details
This is somewhat in limbo? It exists in the API, and there was a demo server for it, which housed the build override for this, but has since been deleted. The changes have been tested, however.

### CODE REVIEW

File: DSharpPlus/Entities/Interaction/Application/DiscordApplicationCommandOption.cs [hunk 1]
```diff
@@ -86,14 +92,16 @@ public sealed class DiscordApplicationCommandOption
         /// <param name="choices">The optional choice selection for this parameter.</param>
         /// <param name="options">The optional subcommands for this parameter.</param>
         /// <param name="channelTypes">The channel types to be restricted to for this parameter, if of type <see cref="ApplicationCommandOptionType.Channel"/>.</param>
-        public DiscordApplicationCommandOption(string name, string description, ApplicationCommandOptionType type, bool? required = null, IEnumerable<DiscordApplicationCommandOptionChoice> choices = null, IEnumerable<DiscordApplicationCommandOption> options = null, IEnumerable<ChannelType> channelTypes = null)
+        public DiscordApplicationCommandOption(string name, string description, ApplicationCommandOptionType type, bool? required = null, bool? autocomplete = null, IEnumerable<DiscordApplicationCommandOptionChoice> choices = null, IEnumerable<DiscordApplicationCommandOption> options = null, IEnumerable<ChannelType> channelTypes = null)
```
Comment (IDoEverything) [inline DSharpPlus/Entities/Interaction/Application/DiscordApplicationCommandOption.cs:95]: This should be added at the end

Comment (VelvetThePanda) [reply DSharpPlus/Entities/Interaction/Application/DiscordApplicationCommandOption.cs:95]: Will do in a minute.

**Comment by VelvetThePanda** [comment]
Also update, there was a PR opened despite the only documentation being in a notion notebook, so I'd assume this is at least subject to change slightly, but discord/discord-api-docs#3849

File: DSharpPlus/Entities/Interaction/Application/DiscordAutoCompleteChoice.cs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        /// Gets the name of this option which will be presented to the user.
+        /// </summary>
+        [JsonProperty("name")]
+        public string Name { get; internal set; }
+
+        /// <summary>
+        /// Gets the value of this option.
+        /// </summary>
+        [JsonProperty("value")]
+        public string Value { get; internal set; }
```
Comment (IDoEverything) [inline DSharpPlus/Entities/Interaction/Application/DiscordAutoCompleteChoice.cs:43]: Autcomplete works for both string and integer option types, so this should be `object` and check for `string` and `long` in the constructor.

Comment (VelvetThePanda) [reply DSharpPlus/Entities/Interaction/Application/DiscordAutoCompleteChoice.cs:43]: I was aware of this prior to opening the PR (as it's been question on ddevs), but wasn't sure if I should go that route. I suppose JSON.NET could potentially break with the exception of `Cannot convert System.Int32 into System.String` or the likes.

Comment (Emzi0767) [reply DSharpPlus/Entities/Interaction/Application/DiscordAutoCompleteChoice.cs:43]: JSON.NET will actually swallow this fine - it's the only serializer to do so.

Comment (VelvetThePanda) [reply DSharpPlus/Entities/Interaction/Application/DiscordAutoCompleteChoice.cs:43]: I'll be fixing this momentarily.

**Review by IDoEverything — CHANGES_REQUESTED** [review]

File: DSharpPlus/Entities/Interaction/Application/DiscordAutoCompleteChoice.cs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        public object Value { get; internal set; }
+
+        /// <summary>
+        /// Creates a new instance of <see cref="DiscordAutoCompleteChoice"/>.
+        /// </summary>
+        /// <param name="name">The name of this option, which will be presented to the user.</param>
+        /// <param name="value">The value of this option.</param>
+        public DiscordAutoCompleteChoice(string name, object value)
+        {
+            if (value is not (string or int))
```
[the remark this answers is not in the document]
Comment (VelvetThePanda) [reply DSharpPlus/Entities/Interaction/Application/DiscordAutoCompleteChoice.cs:52]: Is that supported?

[the remark this answers is not in the document]
Comment (IDoEverything) [reply DSharpPlus/Entities/Interaction/Application/DiscordAutoCompleteChoice.cs:52]: Discord allows numbers over the `int` limit so yes

File: DSharpPlus/Entities/Interaction/Application/DiscordApplicationCommandOption.cs [hunk 4]
```diff
@@ -86,14 +92,16 @@ public sealed class DiscordApplicationCommandOption
         /// <param name="choices">The optional choice selection for this parameter.</param>
         /// <param name="options">The optional subcommands for this parameter.</param>
         /// <param name="channelTypes">The channel types to be restricted to for this parameter, if of type <see cref="ApplicationCommandOptionType.Channel"/>.</param>
-        public DiscordApplicationCommandOption(string name, string description, ApplicationCommandOptionType type, bool? required = null, IEnumerable<DiscordApplicationCommandOptionChoice> choices = null, IEnumerable<DiscordApplicationCommandOption> options = null, IEnumerable<ChannelType> channelTypes = null)
+        public DiscordApplicationCommandOption(string name, string description, ApplicationCommandOptionType type, bool? required = null, IEnumerable<DiscordApplicationCommandOptionChoice> choices = null, IEnumerable<DiscordApplicationCommandOption> options = null, IEnumerable<ChannelType> channelTypes = null, bool? autocomplete = null)
```
Comment (IDoEverything) [inline DSharpPlus/Entities/Interaction/Application/DiscordApplicationCommandOption.cs:95]: Missing xml documentation for the parameter

**Pull request merged**